app/src/main/java/com/github/se/project/viewmodels/AuthenticationViewModel.kt

```diff
@@ -0,0 +1,140 @@
+package com.github.se.project.viewmodels
+
+import android.content.Context
+import android.widget.Toast
+import androidx.credentials.CredentialManager
+import androidx.credentials.CustomCredential
+import androidx.credentials.GetCredentialRequest
+import androidx.credentials.exceptions.GetCredentialCancellationException
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.github.se.project.R
+import com.google.android.libraries.identity.googleid.GetGoogleIdOption
+import com.google.android.libraries.identity.googleid.GoogleIdTokenCredential
+import com.google.firebase.auth.AuthResult
+import com.google.firebase.auth.FirebaseAuth
+import com.google.firebase.auth.GoogleAuthProvider
+import java.security.MessageDigest
+import java.util.UUID
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.tasks.await
+
+class AuthenticationViewModel : ViewModel() {
+
+  val userId = MutableLiveData<String>(null)
+
+  // Function to handle Google Sign-In
+  fun handleGoogleSignIn(context: Context, onSuccess: () -> Unit) {
+    viewModelScope.launch {
+      // Collect the result of the Google Sign-In process
+      googleSignIn(context).collect { result ->
+        result.fold( // It allows you to specify actions for both success and failure cases of the
+            // operation, making it easy to manage the different outcomes.
+            onSuccess = { authResult ->
+              // Handle successful sign-in
+              val currentUser = authResult.user
+              if (currentUser != null) {
+                userId.value = currentUser.uid
+
+                Toast.makeText(context, "Sign in successful", Toast.LENGTH_LONG).show()
+
+                onSuccess()
+              }
+            },
+            onFailure = { e ->
+              // Handle sign-in error
+              Toast.makeText(context, "Authentication error: ${e.message}", Toast.LENGTH_LONG)
+                  .show()
+            })
+      }
+    }
+  }
+
+  // Function to perform Google Sign-In and return a Flow of AuthResult
+  private suspend fun googleSignIn(context: Context): Flow<Result<AuthResult>> {
+    // Initialize Firebase Auth instance
+    val firebaseAuth = FirebaseAuth.getInstance()
+
+    // Return a Flow that emits the result of the Google Sign-In process
+    return callbackFlow {
```

Comment:
We could eventually use a StateFlow instead of a callbackFlow for state management but we could leave it like that and adjust it once we test it tomorrow with the other screens